apm-lambda-extension/apmproxy/apmserver.go

```diff
@@ -210,11 +222,46 @@ func (c *Client) ComputeGracePeriod() time.Duration {
 
 // EnqueueAPMData adds a AgentData struct to the agent data channel, effectively queueing for a send
 // to the APM server.
-func (c *Client) EnqueueAPMData(agentData AgentData) {
+func (c *Client) EnqueueAPMData(agentData AgentData) bool {
 	select {
 	case c.DataChannel <- agentData:
 		c.logger.Debug("Adding agent data to buffer to be sent to apm server")
+		return true
 	default:
 		c.logger.Warn("Channel full: dropping a subset of agent data")
+		return false
 	}
 }
+
+// ShouldFlush returns true if the client should flush APM data after processing the event.
+func (c *Client) ShouldFlush() bool {
+	return c.sendStrategy == SyncFlush || c.hasPendingFlush()
+}
+
+// WaitForFlush returns a channel that is closed if the client has received a signal to flush
+// the buffered APM data.
+func (c *Client) WaitForFlush(ctx context.Context) <-chan struct{} {
+	ch := make(chan struct{})
+	go func() {
+		t := time.NewTimer(10 * time.Millisecond)
```

Comment:
This approach solves the latency issue, so it's definitely better - but I am concerned about busy-waiting causing avoidable CPU overhead. e.g. imagine if the Lambda invocation runs for seconds. I won't block the PR on this as this is better than panicking, but we should follow up on this ASAP.

I wonder if we might be better off with a control loop like in the Go agent, which always runs in the background and receives commands to flush, like: https://github.com/elastic/apm-agent-go/blob/83d9a9518d5e956d04e750a3bae5cbd0d58f682a/tracer.go#L589-L603. That's probably too much of a change for now though.

This would also probably all be simpler if we could pair the `?flushed` requests up with specific Lambda invocation request IDs.

Reply:
I think that's a valid concern, I've removed the busy loop and the code is entirely relying on channels now. It is similar to the original approach before this PR but should handle multiple `flushed=true` requests and avoid race conditions

Reply:
Thanks! I _think_ it's safe now. I also think we should look at reworking the flushing/concurrency in the extension in the not too distant future.